### PR TITLE
fix stdlib type annotation for eval and Pathname#sub

### PR DESCRIPTION
**Pull request by SolaWing**
Pathname sub's signature should be same with String.sub.
eval binding is optional

### CODE REVIEW

**Comment by orangethunder** [comment]
+1 for the eval change. This valid Ruby code is currently rejected by Sorbet:

```
eval "1", nil, "filename", 1
```

**Comment by DarkDimius** [comment]
Kicking off CI, thank you for contribution!

File: rbi/stdlib/pathname.rbi [hunk 1]
```diff
@@ -1040,7 +1040,7 @@ class Pathname < Object
     params(
         p1: String,
     )
-    .returns(String)
+    .returns(Pathname)
```
Comment (DarkDimius) [inline rbi/stdlib/pathname.rbi]: Are you sure it never returns String? Whoever wrote that felt like it sometimes does

Comment (Morriar) [reply rbi/stdlib/pathname.rbi]: Looking at the implementation:

```c
return rb_class_new_instance(1, &str, rb_obj_class(self));
```

It seems to always be returning a `Pathname`.

File: rbi/core/kernel.rbi [hunk 2]
```diff
@@ -223,9 +223,9 @@ module Kernel
   sig do
     params(
         arg0: String,
-        arg1: Binding,
-        filename: String,
-        lineno: Integer,
+        arg1: T.nilable(Binding),
+        filename: T.nilable(String),
```
Comment (orangethunder) [inline rbi/core/kernel.rbi]: I think your original sig was correct (if I recall correctly).

Keep in mind that the only place an _explicit_ nil is allowed is for `arg1`. eval raises an exception when explicit nils are passed for `filename` or `lineno`, but works if they are omitted entirely - sorbet also type-checks that case correctly due the presence of default values (I'm assuming). In other words, I think the sig should only be making `arg` nilable, not the rest. 

In summary, Ruby allows these:

`
eval ":a"
eval ":a", nil
eval ":a", nil, "filename.rb", 123
`

But these are not allowed by Ruby:

`
eval ":a", nil, nil, 12
eval ":a", nil, nil, nil
eval ":a", nil, 12, nil
`

Changing the PR to only allow `arg` to be nilable should allow Sorbet to match Ruby, unless I'm missing something.

Comment (SolaWing) [reply rbi/core/kernel.rbi]: you are right, the filename can be omit, but must be string. I repush it

**Comment by elliottt** [comment]
Thank you!
